Multiple key support

J-Rojas opened the pull request:
Addresses #6. The API changes to use one configuration object to keep the implementation simple.

Also addresses fix for #8

J-Rojas committed:
Support multiple key presses
Support modules

J-Rojas committed:
Add "src" for module support

J-Rojas committed:
Fix bug where closure values were being permentantly overridden in event handler.
Ensure if and only if check for modifier keys. This prevents mutliple key events that support intersecting modified keys sets (eg ['shiftKey', 'altKey'], ['altKey']) from being triggerred with the same key sequence.

J-Rojas committed:
Fix error with preventDefaults if array is used for multiple key events

lupas commented on src/index.vue:
```diff
@@ -1,42 +1,58 @@
   },
   mounted() {
-    window.addEventListener(this.event, this.emitEvent);
+    this._keyListeners = []    
```
Please don't use _ for variables in vue. Doesn't do much difference here in this case, but I wanna avoid it anyway since it's bad practice.
See [documentation](https://vuejs.org/v2/api/#data).

J-Rojas replied:
The 'bad practice' you refer to only applies to reactive items in the 'data' field, not internal component properties that should not be accessed externally. So would you rather have no idoimatic difference between non-reactive and reactive properties? The keyListeners are not meant to be accessed in anyway except in the implementation.

lupas reviewed (changes requested):
Hey, thanks for the PR!

It's cool that your PR allows for multiple key support and fixes some modifier issues, however I don't want to change the usage of the module for that if not necessary.

Most use cases will only use **once key** (or tops with modifiers), so I don't want people to have to input an array of configs in that simple case. Imho that reduces the DX of the module a lot.

A solution like following would be great:

- If a user only wants single-key (standard case), the module can be used as is (with the current props).
- If a user wants multiple key support, the user can enter a multi-key-config (the single-key props are not needed)

Does that make sense to you?

Thanks a lot, would be great to have that in the module!
(and sorry for the late response, quite busy recently)

J-Rojas committed:
Support single key API and config object (multiple key) for backwards compatibility.

J-Rojas commented:
The requested changes have been submitted. If "config" prop is supplied, it will be exclusively used, otherwise the single key properties are used (keycode, event, preventDefault, modifiers). The underlying event implementation remains the same to handle both cases.

lupas commented:
Hey @J-Rojas 

Sorry for the late response, I was super busy the last few days.

I really liked what you did with https://github.com/lupas/vue-keypress/issues/8, that is great and works well!

But I don't think the approach with the `:config` the way you implemented it is very convenient. First of all, it is not really straightforward to understand  with these nested arrays. Furthermore, it seems very cumbersome to maintain for such a simple component, and in my opinion it makes the usage of the module just much more complicated than it has to be.

If you want to track multiple key events, why not just do it like this?
```vue
<template>
  <div>
    <Keypress
      v-for="keypressEvent of keypressEvents"
      :key="keypressEvent.event"
      :key-code="keypressEvent.keyCode"
      :event="keypressEvent.event"
      :modifiers="keypressEvent.modifiers"
      @pressed="keypressEvent.onPressed"
    />
  </div>
</template>

<script>
export default {
  components: {
    Keypress: () => import('vue-keypress'),
  },
  data() {
    return {
      keypressEvents: [
        {
          event: 'keyup',
          keyCode: 37,
          modifiers: ['altKey'],
          onPressed: this.keyupHandler,
        },
        {
          event: 'keydown',
          keyCode: 37,
          modifiers: ['shiftKey'],
          onPressed: this.keydownHandler,
        },
      ],
    }
  },
  methods: {
    keyupHandler(response) {
      console.log('A', response)
    },
    keydownHandler(response) {
      console.log('B', response)
    },
  },
}
</script>
```

Performance wise, this is imho pretty much the same than if we do it within the module, but it already gets rid of one level of complexity.

Now if we want to register multiple keys/modifier combinations per event, we could do it like this:

```vue
<template>
  <div>
    <Keypress
      v-for="keypressEvent of keypressEvents"
      :key="keypressEvent.event"
      :event="keypressEvent.event"
      :multiple="keypressEvent.multiple"
      @pressed="keypressEvent.onPressed"
    />
  </div>
</template>

<script>
export default {
  components: {
    Keypress: () => import('vue-keypress'),
  },
  data() {
    return {
      keypressEvents: [
        {
          event: 'keyup',
          onPressed: this.keyupHandler,
          multiple: [
            {
              keyCode: 37,
              modifiers: ['altKey'],
            },
            {
              keyCode: 57,
              modifiers: ['metaKex'],
            },
          ],
        },
        {
          event: 'keydown',
          onPressed: this.keydownHandler,
          multiple: [
            {
              keyCode: 37,
              modifiers: ['altKey'],
            },
            {
              keyCode: 41,
              modifiers: ['shiftKey'],
            },
          ],
        },
      ],
    }
  },
  methods: {
    keyupHandler(response) {
      console.log('A', response)
    },
    keydownHandler(response) {
      console.log('B', response)
    },
  },
}
</script>
```

The "multiple"  is basically your  "config", just a bit more verbose.

Does that make sense?

In my opinion this would be a more clean approach and would much easier to understand. 

What do you think?

lupas commented on src/index.vue:
```diff
@@ -15,28 +18,59 @@ export default {
-            if (!event[modifier]) return
+    emitEvent(keyCodes, modifiers, preventDefaults) {
+      return (e) => {
+        const index = keyCodes.indexOf(event.keyCode)
```
This breaks in single-key mode because `keyCodes` is not an array.

J-Rojas replied:
Thanks, should be fixed now.

lupas reviewed (changes requested):
see other comment

J-Rojas committed:
Fix single keyCode case

lupas commented:
Did you see my comment above?

J-Rojas commented:
Regarding your comments:

I am using my branch to to track multiple keypresses in my app, at least 6 to start. The motivation for refactoring was to reduce the overhead of maintaining multiple Vue components and multiple key handlers on the DOM. 

I've shown in the my original post how you could halve the number of key handlers using the 'config' approach, all with only one Vue component. Using multiple components may not be a big performance drain - but you'd have to profile it.

Your suggestion is a good explicitly declarative alternative, but this is also fairly verbose. You can make that work without this PR.

I think having a verbose and concise configuration would be a good compromise. This implementation offers the beginning of this to experiment with while supporting the original API. You can decide not to expose the 'config' approach in the documentation for now if you think it needs further discussion, but for the purpose of moving forward with this PR, I've done what I set out to achieve with this. I may decide to extract this as a pure native JS component instead if I don't see too much benefit in reactivity for key management or need a more explicit API.

J-Rojas commented:
Let me know if decide to move forward, otherwise I can close the PR.

J-Rojas committed:
Use while loop to support using only one keyhandler config, if the user choose to do so.
Benefit: debugging this in the console is more straight forward with only one entry point instead of multiple entry points for each key combination.

J-Rojas commented:
Added a while loop to allow having only one key handler for the entire window possible while supporting multiple key combinations. It can become annoying to have to step through multiple handlers if multiple keys are registered while attempting to debug. This can become increasingly important if this component is not solely used at the top level, but across multiple parts of the app.
